### PR TITLE
ref(backup): Convert InstanceID to NamedTuple

### PR DESCRIPTION
**Pull request by azaslavsky**
This is a bit unfortunate, as this requires the `model` to be passed as an (unparsed) `str`, rather than as an already-parsed `NormalizedModelName`, ultimately making the class easier to misuse. However, `InstanceID`s will now need to be sent over RPC, and no other solution I tried for this that retained the type-unwrapping constructor was satisfactory:

- I considered using Pydantic's `BaseModel` or `dataclass` base classes, but opted against this, as I would like to avoid Pydantic types leaking out of the RPC-specific modules they mostly live in today.
- I tried to convince Pydantic to properly encode/decode the existing native dataclass, but ran across https://github.com/pydantic/pydantic/issues/2531, which won't allow custom encoding/decoding of arbitrary types in Pydantic v1.
- I also messed with some other inheritance schemes, but in the end decided converting to a `NamedTuple` and dropping the type unwrapping was probably the best solution for now.

Issue: getsentry/team-ospo#196

### CODE REVIEW

**Comment by azaslavsky** [comment]
@hubertdeng123 this basically undoes a small bit of the refactor I did in https://github.com/getsentry/sentry/pull/56669, due to newly discovered incompatibilities between `@dataclass` and our Pydantic-based RPC system that I did not foresee when submitting that PR. Sorry for the churn!

File: src/sentry/backup/findings.py [hunk 1]
```diff
@@ -19,21 +18,30 @@ class InstanceID:
...
     # can use the ordinal as a unique identifier.
-    ordinal: int | None = None
-
-    def __init__(self, model: NormalizedModelName, ordinal: Optional[int] = None):
-        self.model = str(model)
-        self.ordinal = ordinal
-
-    def __hash__(self):
-        return hash((self.model, self.ordinal))
+    ordinal: Optional[int] = None
```
Comment (hubertdeng123) [inline src/sentry/backup/findings.py:21]: 👏 standardized typing

File: src/sentry/backup/findings.py [hunk 2]
```diff
@@ -19,21 +18,30 @@ class InstanceID:
...
 
     def pretty(self) -> str:
         out = f"InstanceID(model: {self.model!r}"
         if self.ordinal:
             out += f", ordinal: {self.ordinal}"
         return out + ")"
 
+    @classmethod
+    def encode(_, id: InstanceID) -> dict:
```
Comment (hubertdeng123) [inline src/sentry/backup/findings.py:30]: I don't see the `encode` and `decode` functions used anywhere. What is it used for?

Comment (azaslavsky) [reply src/sentry/backup/findings.py:30]: Good catch - these are vestigial from another approach I had taken. They've been removed.

**Review by hubertdeng123 — APPROVED** [review]

**Pull request merged**